docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,243 @@
+
+There is discussion about changing the behaviour of `preflight`, considering that preflights and support-bundles utilize the same specs - the same collectors and analyzers - and only differ in what is returned to the user post-analysis.  For this design proposal, `support-bundle` and `preflight` may be condensed into a single `troubleshoot` binary.
+
+For the purposes of this design we'll talk about troubleshoot as a standalone binary. It will also be available, and likley most commonly installed as, a krew plugin for kubectl.
+
+## Non-Goals
+
+## Background
+
+## High Level Design
+
+Functions of `support-bundle`, `preflight`, `analyze`, `redact`, and `sbctl` binaries/tools should be rolled together into a single `troubleshoot` CLI that can perform all necessary functions of the same.
+
+`troubleshoot` CLI should be able to report on the version that is installed in the CLI, and any support bundle generated with `troubleshoot` should report its build/version inside the archive it generates.
+
+## Detailed Design
+
+In the interest of being able to work on this quickly without breaking existing use-cases, a new `troubleshoot` command should be created. Utilizing cobra and viper best practices from the cobra.dev docs.
+
+### sbctl → troubleshoot inspect
+
+sbctl should be migrated to the troubleshoot repository in a "lift and shift" operation to start with.
+
+It should continue to be built as a standalone binary to enable continued use until a stable replacement exists. as such it should me integrated in a way that preserves the existing `cmd` and it's package requirements.
+
+Once integrated into the repository, we can begin using functions from it's packages from the from the new `troubleshoot` command to enable the `inspect` subcommand.
+
+### Public APIs
+
+The existing package entrypoint we use from the support-bundle command is `supportbundle.CollectSupportBundleFromSpec()`
+which is an end-to-end function that runs collectors, redactors and then analysis.
+
+The Consolidated cli should be responsible for handling the end-to-end running of this process in the `cmd` leaving the importable packages free to focus on defining the individual collect, redact analyze steps.
+
+To this end we should create a new `pkg` that targets the functionality provided by the `collect` and `analyze` packages such as `collect.runHostCollectors()` and provide a stable API to be used by the CLI and other projects. At the same time we can unpublish functionality that should not be exposed (such as the individual collector and analyser functions) and mark code for deprecation.
+
+This new package should be kept as minimal as possible. and serve only as an interface to private functions in the other packages.
+
+Once the stable API is ready we can instruct projects like kurl to target that and work on removing code marked for deprecation.
```

Comment:
_Implementation detail_

All internal logic should be moved from the `pkg` package to an [`internal` package](https://go.dev/doc/go1.4#internalpackages).

For reference of how this would look like, please take a look at https://github.com/derailed/k9s/tree/master/internal project that uses this pattern extensively.

Reply:
In k9s, this really only changes the import path rather than much else.  The package files still start with, e.g. `package client`.  Kubernetes itself doesn't use this pattern though there are many packages not intended to be imported by other projects.  Is there something about using the 'internal' directory other than the name making it clearer as to the intent of these packages?  I'd rather avoid complexity, but not sure if using 'internal' avoids complexity or creates it.

Reply:
> Is there something about using the 'internal' directory other than the name making it clearer as to the intent of these packages? I'd rather avoid complexity, but not sure if using 'internal' avoids complexity or creates it.

Its [golang's way of enforcing](https://go.dev/doc/go1.4#internalpackages) (at a compiler level) the boundary of public vs private packages. In troubleshoot for example, `github.com/replicatedhq/troubleshoot/internal/traces` package cannot be imported by any other project besides troubleshoot.